Modify default compaction planner opts to include internal type

dlmarion opened the pull request:
Added type property to the planner configuration to resolve
the CompactionIT failures seen in #2137

ctubbsii commented on server/tserver/src/main/java/org/apache/accumulo/tserver/compactions/CompactionManager.java:
```diff
@@ -145,10 +145,11 @@ private long getDefaultThroughput(AccumuloConfiguration aconf) {
+                  DefaultCompactionPlanner.class.getName(),
                   defaultServicePrefix + "planner.opts.executors",
-                  "[{'name':'deprecated','numThreads':" + numThreads + "}]");
+                  "[{'name':'deprecated', 'type':'internal', 'numThreads':" + numThreads + "}]");
```
Rather than make this change, I think we should default to inferring an internal type when it is unspecified. Explicitly specifying `internal` should not be required.

dlmarion replied:
Implemented in 58673f8

keith-turner commented:
Should `type':'internal'` be removed from the defaults in Property.java?

dlmarion commented:
Personally, I like it there as it's explicit that you (as an administrator) are making a decision between internal and external. There is no harm in removing it or leaving it.

keith-turner commented:
> Personally, I like it there as it's explicit that you

Yeah that seems good.

ctubbsii commented on core/src/main/java/org/apache/accumulo/core/spi/compaction/DefaultCompactionPlanner.java:
```diff
@@ -161,8 +161,13 @@ public void init(InitParameters params) {
+      // If not supplied, GSON will leave type null. Default to internal
+      if (null == executorConfig.type) {
+        executorConfig.type = "internal";
+      } else if (!("internal".equals(executorConfig.type)
+          || "external".equals(executorConfig.type))) {
```
I've noticed that Eclipse has some settings that suggest expanding negation of compound boolean expressions. I like these suggestions, because I think it's easier for the human mind to interpret the compound of separate negations ("it's not a pig and it's not a dog") than the negation of a compound ("it's not the case that it's a pig or a dog").

So, I suggest:

```java
      } else if (!"internal".equals(executorConfig.type)
          && !"external".equals(executorConfig.type)) {
```

Or, maybe something with set inclusion instead:

```suggestion
      } else if (!Set.of("internal", "external").contains(executorConfig.type)) {
```

Or, just remove this check entirely, since the default case in the switch statement just below it already handles this.

dlmarion replied:
Good call on the removal, it is handled below that. I'm removing this.

dlmarion replied:
Removed in f271227

ctubbsii commented:
> Personally, I like it there as it's explicit that you (as an administrator) are making a decision between internal and external. There is no harm in removing it or leaving it.

Having it explicit in the default config is good, as is allowing it to be omitted. However, we should have test coverage of the case where it is omitted.

dlmarion commented:
Regarding test coverage, we have it, it's why CompactionIT was failing.

ctubbsii commented on core/src/main/java/org/apache/accumulo/core/spi/compaction/DefaultCompactionPlanner.java:
```diff
@@ -161,8 +161,11 @@ public void init(InitParameters params) {
 
-      Objects.requireNonNull(executorConfig.type,
-          "'type' is a required and must be 'internal' or 'external'");
+      // If not supplied, GSON will leave type null. Default to internal
+      if (null == executorConfig.type) {
+        executorConfig.type = "internal";
+      }
```
Not long ago, somebody swept through the code to standardize on `variable == null` rather than `null == variable`. It's not too important, but in order to not drift too far from that person's previous efforts to standardize things, we should probably strive to maintain that same style. I don't know if we can enforce this one in the style rules, though.

```suggestion
      // If not supplied, GSON will leave type null. Default to internal
      if (executorConfig.type == null) {
        executorConfig.type = "internal";
      }
```

ctubbsii commented:
> Regarding test coverage, we have it, it's why CompactionIT was failing.

My understanding was that the test was failing with the default config, which you changed to be explicit. I didn't see a test if the user overrides the default config with the type set to null.

dlmarion commented:
I reverted the explicit addition of the type in the default configuration in 58673f8. It's now added if the type is missing in the latest version of code.

ctubbsii commented:
> I reverted the explicit addition of the type in the default configuration in [58673f8](https://github.com/apache/accumulo/commit/58673f826da46db05c487daf75b2ae4dce14d816). It's now added if the type is missing in the latest version of code.

Oh, okay. I thought you were going to keep the explicit type, based on the conversation above, and because I saw that the file remained changed after you reverted that, but that is just a whitespace diff and not the original change. Either way, this is good to merge.
